packages/oauth-js/src/client.ts

```diff
@@ -0,0 +1,361 @@
+    saveSessionData(sessionData, this.options);
+
+    const storageAPI = getStorageAPI(this.options.useCookie);
+    const returnURL = storageAPI.getItem(LAST_PATH_KEY) || "/";
+
+    const appState = {
+      userData: user,
+      returnPath: returnURL,
+      authState: storageAPI.getItem(STATE_DATA_KEY),
+    };
+
+    storageAPI.removeItem(LAST_PATH_KEY);
+
+    if (this.options.useCookie) {
+      setTokenCookies(user, this.options);
+    }
+
+    this.startTokenWatch();
+
+    if (this.config.onLogin) {
+      this.config.onLogin(appState);
+    }
+  }
+
+  private async exchange() {
+    const queryString = window.location.search;
+    const urlParams = new URLSearchParams(queryString);
+    const savedState = this.storage.getItem(STATE_DATA_KEY);
+    const codeVerifier = this.storage.getItem(VERIFIER_DATA_KEY);
+
+    const state = urlParams.get("state");
+    const code = urlParams.get("code");
+
+    this.authState = "UPDATING";
+
+    if (!state || !code) {
+      const msg = "Missing required parameters";
+      this.authState = "ERROR";
+    } else if (state !== savedState || savedState === "") {
+      const msg = "Invalid session state";
+      this.authState = "ERROR";
+    } else {
+      try {
+        const response = await this.post("token", {
+          client_id: this.config.clientId,
+          redirect_uri: this.config.callbackUri,
+          grant_type: "authorization_code",
+          code_verifier: codeVerifier,
+          code,
+        });
+
+        if (response.ok) {
+          this.authState = "AUTHENTICATED";
+          this.startTokenWatch();
+          this.processResponse(response);
+        }
+      } catch (err) {
+        this.authState = "ERROR";
+      }
+    }
+  }
+
+  private shouldTokenRefresh() {
+    switch (this.authState) {
+      case "INIT":
+      case "UPDATING":
+        return true;
+      case "ERROR":
+        return false;
+      case "AUTHENTICATED":
+        const [token, expireTime] = getSessionTokenValues(this.options);
+
+        if (!token) {
+          return true;
+        }
+
+        return isTokenExpiring(expireTime);
+    }
+  }
+
+  getUrl(endpoint: string): string {
+    const protocol = this.config.domain.match(/^local\.?host(:\d{2,5})?$/)
+      ? "http"
+      : "https";
+
+    return `${protocol}://${this.config.domain}/${API_VERSION}/${endpoint}`;
+  }
+
+  getHeaders(): any {
+    const headers = {
+      headers: {
+        "Content-Type": "application/json",
+      },
+    };
+
+    return headers;
+  }
+
+  private async get(endpoint: string, params?: any) {
+    const url = this.getUrl(endpoint);
+    const query = new URLSearchParams(params);
+    return fetch(`${url}?${query}`, {
+      method: "GET",
+      headers: {
+        Authorization: `Basic ${this.config.clientId}`,
+      },
+    });
+  }
+
+  private async post(endpoint: string, payload: any) {
+    const data = new URLSearchParams(payload);
+    const credentials = btoa(this.config.clientId + ":");
```

Comment:
when is clientSecret added?

Reply:
In the PCKE flow the client secret isn't needed. The basic auth header shouldn't be needed here either, this is to work around a bug in the backend for now. Passing the client_id in the body is support to work.